File: packages/sdk-node/src/runtime.ts
import { resolveConfig, type ChoboConfigInput, type ChoboConfig } from "./config.js";
import { Transport, type TransportStats } from "./transport.js";
import type { ChoboEvent } from "./event.js";

interface RuntimeState {
  transport: Transport | null;
  config: ChoboConfig | null;
}

// Hang the singleton off globalThis so the ESM and CJS builds share one instance
// (dual-package hazard mitigation — see docs/research/2026-06-24-node-sdk-grounding.md §4).
const KEY = Symbol.for("chobo.runtime.v1");
function state(): RuntimeState {
  const g = globalThis as unknown as Record<symbol, RuntimeState | undefined>;
  if (!g[KEY]) g[KEY] = { transport: null, config: null };
  return g[KEY] as RuntimeState;
}

export function init(input: ChoboConfigInput): ChoboConfig {
  const s = state();
  if (s.transport) void s.transport.shutdown();
  const cfg = resolveConfig(input);
  s.config = cfg;
  s.transport = new Transport(cfg);
  return cfg;
}

export function emit(event: ChoboEvent): void {
  state().transport?.enqueue(event);
}

export function flush(): Promise<void> {
  return state().transport?.flush() ?? Promise.resolve();
}

export async function shutdown(): Promise<void> {
  const s = state();
  await (s.transport?.shutdown() ?? Promise.resolve());
  s.transport = null;
  s.config = null;
}

export function getStats(): Partial<TransportStats> {
  const t = state().transport;
  return t ? { ...t.stats } : {};
}

export function getConfig(): ChoboConfig | null {
  return state().config;
}

/** Test helper: tear down the active transport. */
export function reset(): Promise<void> {
  return shutdown();
}

File: packages/sdk-python/src/chobo/_runtime.py
"""Global SDK runtime: holds the active Config + Transport, wires lifecycle helpers."""
from .config import Config
from .transport import Transport

_transport = None
_config = None


def init(ingest_url, service, account=None, **kwargs):
    global _transport, _config
    if _transport is not None:
        reset()
    _config = Config(ingest_url=ingest_url, service=service, account=account, **kwargs)
    _transport = Transport(
        ingest_url=_config.ingest_url,
        queue_maxsize=_config.queue_maxsize,
        batch_max=_config.batch_max,
        flush_interval=_config.flush_interval,
        spool_dir=_config.spool_dir,
        max_spool_bytes=_config.max_spool_bytes,
        timeout=_config.timeout,
        ingest_secret=_config.ingest_secret,
    )
    return _config


def get_config():
    return _config


def emit(event):
    if _transport is None:
        return
    _transport.enqueue(event)


def flush(timeout=5.0):
    if _transport is None:
        return True
    return _transport.flush(timeout=timeout)


def shutdown(timeout=5.0):
    if _transport is not None:
        _transport.shutdown(timeout=timeout)


def get_stats():
    return dict(_transport.stats) if _transport is not None else {}


def reset():
    """Test helper: tear down the active transport."""
    global _transport, _config
    if _transport is not None:
        _transport.shutdown(timeout=2)
    _transport = None
    _config = None
